heartbeat/scheduler/timerqueue/heap.go

```diff
@@ -0,0 +1,51 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timerqueue
+
+// timerHeap is the internal type that implements container/heap
+type timerHeap []*TimerTask
```

Comment:
How about `timerHeap []TimerTask`? Right now another object is creating and scheduling the task by passing a pointer. Nothing prevents other objects to mess with the timestamp.

Reply:
AFAIK slices of pointers are usually faster (see https://stackoverflow.com/questions/27622083/slices-of-structs-vs-slices-of-pointers-to-structs ).

I'm not too worried about mutability given that we're the only users of this API. I'll clarify that the fields should not be changed after being scheduled.